### PR TITLE
Modified package to now use class based approach

### PR DESCRIPTION
**Pull request by gavinhenderson**
Here are the changes proposed in #9. They are by no means complete just wanted to get some opinions on the changes made so far before I put more time into tidying the code and adding more tests.

### CODE REVIEW

**Comment by simonepri** [comment]
Hi @gavinhenderson I'll review this asap.
Thank you for the time you spent contributing!

File: index.js [hunk 1]
```diff
@@ -16,7 +25,7 @@ const queue = [];
  * @param {Function} algorithm.identifiers A function that returns the list of
  * identifiers that this password hashing algorithm is able to generate / verify.
  */
-function install(name, algorithm) {
+Upash.prototype.install = function(name, algorithm) {
```
Comment (simonepri) [inline index.js]: This should be private, we want the instance to be immutable

File: index.js [hunk 2]
```diff
@@ -46,54 +55,54 @@ function install(name, algorithm) {
...
+};
 
 /**
  * Uninstalls a password hashing function previously installed.
  * @public
  * @param {string} name The name of the algorithm to uninstall or 'last' to
  * uninstall the last one installed.
  */
-function uninstall(name) {
+Upash.prototype.uninstall = function(name) {
```
Comment (simonepri) [inline index.js]: We should remove this

File: index.js [hunk 3]
```diff
@@ -102,32 +111,32 @@ function list() {
  * @param {string|undefined} name The name of the algorithm to use.
  * @return {Object} The password hashing function object.
  */
-function use(name) {
+Upash.prototype.use = function(name) {
```
Comment (simonepri) [inline index.js:80]: This should set the default for the current instance

**Review by simonepri — CHANGES_REQUESTED** [review]

**Comment by gavinhenderson** [comment]
@simonepri Thats those three changes made now. If you think these changes are okay let me know and I'll add more tests for the new API and I'll improve the way setting default works.

**Comment by gavinhenderson** [comment]
Hey @simonepri , just wondering if there was any more changes you wanted me to make 👍

**Comment by simonepri** [comment]
Hey @gavinhenderson sorry for the long delay, It looks cool.
I'll dive into this during the weekend and I'll let know you know.

Thank you for contributing!

**Comment by gavinhenderson** [comment]
Hey @simonepri, I hope youre well! I noticed there hasn't been any changes on master for 4 months. Is there plans to mainain this project or should I close my pull request?

**Comment by simonepri** [comment]
@gavinhenderson 
No Gavin, I'm super happy with this PR.
I'm working on a refactor for all the `phc-*` packages that `upash` relies on, and I want to finish it before merging this pr.

Sorry for the long delay.

**Comment by indo-dev-0** [comment]
Updates on this? I would like to contribute changes that would benefit from a class based system.

**Comment by gavinhenderson** [comment]
@indo-dev-0 no updates from me

**Comment by simonepri** [comment]
Sorry for the 1-year long delay. (I had an intense year)
I'll try to review and merge this before end of next week.

**Pull request closed without merging**